pkg/validate/volumes.go

```diff
@@ -0,0 +1,53 @@
+// Copyright The Shipwright Contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package validate
+
+import (
+	"fmt"
+
+	build "github.com/shipwright-io/build/pkg/apis/build/v1alpha1"
+	buildv1alpha1 "github.com/shipwright-io/build/pkg/apis/build/v1alpha1"
+)
+
+// BuildVolumes is used to validate volumes in the Build object
+func BuildVolumes(strategyVolumes []buildv1alpha1.BuildStrategyVolume, buildVolumes []buildv1alpha1.BuildVolume) (bool, buildv1alpha1.BuildReason, string) {
+	return validateVolumes(strategyVolumes, buildVolumes)
+}
+
+// BuildRunVolumes is used to validate volumes in the BuildRun object
+func BuildRunVolumes(strategyVolumes []buildv1alpha1.BuildStrategyVolume, buildVolumes []buildv1alpha1.BuildVolume) (bool, string, string) {
+	valid, reason, msg := validateVolumes(strategyVolumes, buildVolumes)
+	return valid, string(reason), msg
+}
+
+// validateBuildVolumes validates build overriding the build strategy volumes. in case it tries
+// to override the non-overridable volume, or volume that does not exist in the strategy, it is
+// good to fail early
+func validateVolumes(strategyVolumes []build.BuildStrategyVolume, buildVolumes []build.BuildVolume) (bool, buildv1alpha1.BuildReason, string) {
+	strategyVolumesMap := toVolumeMap(strategyVolumes)
+
+	for _, buildVolume := range buildVolumes {
+		strategyVolume, ok := strategyVolumesMap[buildVolume.Name]
+		if !ok {
+			return false, buildv1alpha1.UndefinedVolume, fmt.Sprintf("Volume %q is not defined in the Strategy", buildVolume.Name)
+		}
+
+		// nil for overridable is equal to false
+		if !strategyVolume.Overridable {
```

Comment:
I think your comment makes sense. But you changed `Overridable` from `*bool` to `bool`. Recently we changed all optional fields in Builds and BuildRuns to pointers. I think that would apply here as well.

Reply:
I saw some other optional field being non-pointer in kubernetes.
And, since boolean value logically cannot represent anything but true or false (meaning volume is either overridable or not, and cannot be in any other, third logical state), switching to non-pointer makes sense in order to minimize human error while dealing with possible null pointers.
I can switch back though, if needed.